src/pool.ts

```diff
@@ -0,0 +1,729 @@
+  onConnect?: (connection: AwaitConnection) => Promise<void>;
+  connectionFactory?: (options?: ConnectConfig) => Promise<AwaitConnection>;
+}
+
+/**
+ * A connection pool.
+ *
+ * Connection pool can be used to manage a set of connections to the database.
+ * Connections are first acquired from the pool, then used, and then released
+ * back to the pool. Once a connection is released, it's reset to close all
+ * open cursors and other resources *except* prepared statements.
+ * Pools are created by calling :func:`~edgedb.pool.create`.
+ */
+export class Pool {
+  private _closed: boolean;
+  private _closing: boolean;
+  private _queue: LifoQueue<PoolConnectionHolder>;
+  private _holders: PoolConnectionHolder[];
+  private _initialized: boolean;
+  private _initializing: boolean;
+  private _minSize: number;
+  private _maxSize: number;
+  private _onAcquire?: (proxy: PoolConnectionProxy) => Promise<void>;
+  private _onRelease?: (proxy: PoolConnectionProxy) => Promise<void>;
+  private _onConnect?: (connection: AwaitConnection) => Promise<void>;
+  private _connectionFactory: (
+    options?: ConnectConfig
+  ) => Promise<AwaitConnection>;
+  private _generation: number;
+  private _connectOptions: ConnectConfig;
+
+  public get initialized(): boolean {
+    return this._initialized;
+  }
+
+  /** @internal */
+  public get generation(): number {
+    return this._generation;
+  }
+
+  /** @internal */
+  public get holders(): PoolConnectionHolder[] {
+    return this._holders;
+  }
+
+  /** @internal */
+  public get queue(): LifoQueue<PoolConnectionHolder> {
+    return this._queue;
+  }
+
+  public get minSize(): number {
+    return this._minSize;
+  }
+
+  public get maxSize(): number {
+    return this._maxSize;
+  }
+
+  protected constructor(options: PoolOptions) {
+    const {onAcquire, onRelease, onConnect, connectOptions} = options;
+    const minSize =
+      options.minSize === undefined ? DefaultMinPoolSize : options.minSize;
+    const maxSize =
+      options.maxSize === undefined ? DefaultMaxPoolSize : options.maxSize;
+
+    this.validateSizeParameters(minSize, maxSize);
+
+    this._queue = new LifoQueue<PoolConnectionHolder>();
+    this._holders = [];
+    this._initialized = false;
+    this._initializing = false;
+    this._minSize = minSize;
+    this._maxSize = maxSize;
+    this._onAcquire = onAcquire;
+    this._onRelease = onRelease;
+    this._onConnect = onConnect;
+    this._closing = false;
+    this._closed = false;
+    this._generation = 0;
+    this._connectOptions = connectOptions;
+    this._connectionFactory = options.connectionFactory ?? connect;
+  }
+
+  static async create(options: PoolOptions): Promise<Pool> {
+    const pool = new Pool(options);
+    await pool.initialize();
+    return pool;
+  }
+
+  private validateSizeParameters(minSize: number, maxSize: number): void {
+    if (maxSize <= 0) {
+      throw new errors.InvalidValueError(
```

Comment:
Should be `errors.InterfaceError`.

In general, most of exceptions defined in errors (coming from [`edb/api/errors.txt`](https://github.com/edgedb/edgedb/blob/master/edb/api/errors.txt)) are expected to be raised from the EdgeDB *server*.

We have a few errors reserved for *client side* errors, such as errors caused by improper use of our APIs:

```
#### Suggested errors for EdgeDB clients

0x_FF_00_00_00   ClientError
0x_FF_01_00_00   ClientConnectionError

0x_FF_02_00_00   InterfaceError

0x_FF_02_01_00   QueryArgumentError
0x_FF_02_01_01   MissingArgumentError
0x_FF_02_01_02   UnknownArgumentError
0x_FF_02_01_03   InvalidArgumentError

0x_FF_03_00_00   NoDataError
```